rfcs/013 - Datastore Revisited RFC.md

```diff
@@ -0,0 +1,142 @@
+
+```ln
+from @std/app import start, print, exit
+from @std/datastore import namespace, set, getOr
+
+on start {
+  const foo = namespace('foo');
+  foo.set('bar', 5);
+  print(foo.getOr('bar', 1));
+  print(foo.getOr('bar', 1.0));
+  emit exit 0;
+}
+```
+
+not only works, but it produces the following output
+
+```
+5
+0.000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000025
+```
+
+The reason why this is a bug and not a feature is that it's an *unsafe* casting between types, so this
+
+```ln
+from @std/app import start, print, exit
+from @std/datastore import namespace, set, getOr
+
+on start {
+  const foo = namespace('foo');
+  foo.set('bar', 5);
+  print(foo.getOr('bar', 1));
+  print(foo.getOr('bar', ''));
+  emit exit 0;
+}
+```
+
+produces this
+
+```
+thread 'tokio-runtime-worker' panicked at 'range end index 5 out of range for slice of length 0', src/vm/memory.rs:628:29
+note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
+thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: JoinError::Panic(...)', src/vm/event.rs:260:14
+```
+
+a hard crash of the runtime as it tries to access a number as if it was a string. With a stated goal of eliminating as many runtime errors as possible so you can be sure that your backend is going to continue running, having this unsafe casting footgun was absolutely not the intention of the `datastore` library.
+
+There are multiple ways to resolve this such that the actual datastore opcodes are left alone while the compiler guarantees that their usage is safe. For the sake of unblocking datastore usage as fast as possible, we're going to go through a two-step process to get it working, if a bit weird, and then using a new language feature to implement it cleanly. This does mean that the API will be changed twice, but since we are so far away from 1.0, this is considered fine for the adventurous who want to use this standard library that isn't documented publicly, yet.
+
+### Part 1: Minimally-invasive with restored type introspection
+
+We recently removed the ability to get the type of a variable as a string with the `type variableName` syntax, but if we restore that, it should be possible to force type safety by combining the type string with either the namespace or the key string, such that trying to get the same key name with a default value of a different type will find that it doesn't exist and return the provided default.
+
+This will provide zero actual API call changes except to the `has` and `del` functions, which would need to be given an example of the type of data to delete in order to find it. This has the smallest change requirement, though the API is a bit weird.
```

Comment:
I think it's worth noting that this won't be necessary once part 2 is implemented

Reply:
Sure. But I'll note in in the second section.

Reply:
Wait, I did note that here: https://github.com/alantech/alan/pull/378/files#diff-5268a48223be94b011eb9b7b7042590372f2316a618e53850b6f9b0b69190635R95